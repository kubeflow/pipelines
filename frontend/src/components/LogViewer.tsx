/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { List, AutoSizer, ListRowProps } from 'react-virtualized';
import { fontsize, fonts } from '../Css';
import { stylesheet } from 'typestyle';
import { OverscanIndicesGetter } from 'react-virtualized/dist/es/Grid';

const css = stylesheet({
  a: {
    $nest: {
      '&:hover': {
        color: '#bcdeff',
      },
      '&:visited': {
        color: '#bc9fff',
      },
    },
    color: '#96cbfe',
  },
  line: {
    $nest: {
      '&:hover': {
        backgroundColor: '#333',
      },
    },
    display: 'flex',
  },
  number: {
    color: '#999',
    flex: '40px 0 0',
    paddingRight: 10,
    textAlign: 'right',
    userSelect: 'none',
  },
  root: {
    // We cannot easily add padding here without breaking react-virtualized size calculation, for
    // details: https://github.com/bvaughn/react-virtualized/issues/992
    // Specifically, a complex solution was proposed in https://github.com/bvaughn/react-virtualized/issues/992#issuecomment-371145943.
    // We may consider that later.
    backgroundColor: '#222',
    color: '#fff',
    fontFamily: fonts.code,
    fontSize: fontsize.small,
    // This override and listContainerStyleOverride workarounds to allow horizontal scroll.
    // Reference: https://github.com/bvaughn/react-virtualized/issues/1248
    overflow: 'auto !important',
    whiteSpace: 'pre',
  },
});

const listContainerStyleOverride = {
  overflow: 'visible',
};

interface LogViewerProps {
  logLines: string[];
}

// Use the same amount of overscan above and below visible rows.
//
// Why:
// * Default behavior is that when we scroll to one direction, content off
// screen on the other direction is unmounted from browser immediately. This
// caused a bug when selecting lines + scrolling.
// * With new behavior implemented below: we are now overscanning on both
// directions disregard of which direction user is scrolling to, we can ensure
// lines not exceeding maximum overscanRowCount lines off screen are still
// selectable.
const overscanOnBothDirections: OverscanIndicesGetter = ({
  direction, // One of "horizontal" or "vertical"
  cellCount, // Number of rows or columns in the current axis
  scrollDirection, // 1 (forwards) or -1 (backwards)
  overscanCellsCount, // Maximum number of cells to over-render in either direction
  startIndex, // Begin of range of visible cells
  stopIndex, // End of range of visible cells
}) => {
  return {
    overscanStartIndex: Math.max(0, startIndex - overscanCellsCount),
    overscanStopIndex: Math.min(cellCount - 1, stopIndex + overscanCellsCount),
  };
};

interface LogViewerState {
  followNewLogs: boolean;
}

class LogViewer extends React.Component<LogViewerProps, LogViewerState> {
  public state = {
    followNewLogs: true,
  };

  private _rootRef = React.createRef<List>();

  public componentDidMount(): void {
    // Wait until the next frame to scroll to bottom, because doms haven't been
    // rendered when running this.
    setTimeout(() => {
      this._scrollToEnd();
    });
  }

  public componentDidUpdate(): void {
    if (this.state.followNewLogs) {
      this._scrollToEnd();
    }
  }

  public render(): JSX.Element {
    return (
      <AutoSizer>
        {({ height, width }) => (
          <List
            id='logViewer'
            containerStyle={listContainerStyleOverride}
            width={width}
            height={height}
            rowCount={this.props.logLines.length}
            rowHeight={15}
            className={css.root}
            ref={this._rootRef}
            overscanIndicesGetter={overscanOnBothDirections}
            overscanRowCount={
              400 /* make this large, so selecting maximum 400 lines is supported */
            }
            rowRenderer={this._rowRenderer.bind(this)}
            onScroll={this.handleScroll}
          />
        )}
      </AutoSizer>
    );
  }

  private handleScroll = (info: {
    clientHeight: number;
    scrollHeight: number;
    scrollTop: number;
  }) => {
    const offsetTolerance = 20; // pixels
    const isScrolledToBottom =
      info.scrollHeight - info.scrollTop - info.clientHeight <= offsetTolerance;
    if (isScrolledToBottom !== this.state.followNewLogs) {
      this.setState({
        followNewLogs: isScrolledToBottom,
      });
    }
  };

  private _scrollToEnd(): void {
    const root = this._rootRef.current;
    if (root) {
      root.scrollToRow(this.props.logLines.length + 1);
    }
  }

  private _rowRenderer(props: ListRowProps): React.ReactNode {
    const { style, key, index } = props;
    const line = this.props.logLines[index];
    return (
      <div key={key} className={css.line} style={style}>
        <MemoedLogLine index={index} line={line} />
      </div>
    );
  }
}

const LogLine: React.FC<{ index: number; line: string }> = ({ index, line }) => (
  <>
    <span className={css.number} style={getLineStyle(line)}>
      {index + 1}
    </span>
    <span className={css.line} style={getLineStyle(line)}>
      {parseLine(line).map((piece, p) => (
        <span key={p}>{piece}</span>
      ))}
    </span>
  </>
);
// improve performance when rerendering, because we render a lot of logs
const MemoedLogLine = React.memo(LogLine);

function getLineStyle(line: string): React.CSSProperties {
  const lineLowerCase = line.toLowerCase();
  if (lineLowerCase.indexOf('error') > -1 || lineLowerCase.indexOf('fail') > -1) {
    return {
      backgroundColor: '#700000',
      color: 'white',
    };
  } else if (lineLowerCase.indexOf('warn') > -1) {
    return {
      backgroundColor: '#545400',
      color: 'white',
    };
  } else {
    return {};
  }
}

function parseLine(line: string): React.ReactNode[] {
  // Linkify URLs starting with http:// or https://
  // eslint-disable-next-line no-useless-escape
  const urlPattern = /(\b(https?):\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/gim;
  let lastMatch = 0;
  let match = urlPattern.exec(line);
  const nodes = [];
  while (match) {
    // Append all text before URL match
    nodes.push(<span>{line.substr(lastMatch, match.index)}</span>);
    // Append URL via an anchor element
    nodes.push(
      <a href={match[0]} target='_blank' rel='noopener noreferrer' className={css.a}>
        {match[0]}
      </a>,
    );

    lastMatch = match.index + match[0].length;
    match = urlPattern.exec(line);
  }
  // Append all text after final URL
  nodes.push(<span>{line.substr(lastMatch)}</span>);
  return nodes;
}

export default LogViewer;
