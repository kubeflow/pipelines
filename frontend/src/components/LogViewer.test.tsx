/*
 * Copyright 2018 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { mount, shallow } from 'enzyme';
import LogViewer from './LogViewer';

describe('LogViewer', () => {
  it('renders an empty container when no logs passed', () => {
    expect(shallow(<LogViewer logLines={[]} />)).toMatchSnapshot();
  });

  it('renders one log line', () => {
    const logLines = ['first line'];
    const logViewer = new LogViewer({ logLines });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders two log lines', () => {
    const logLines = ['first line', 'second line'];
    const logViewer = new LogViewer({ logLines });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders one long line without breaking', () => {
    const line =
      `Lorem Ipsum is simply dummy text of the printing and typesetting` +
      `industry. Lorem Ipsum has been the industry's standard dummy text ever` +
      `since the 1500s, when an unknown printer took a galley of type and` +
      `scrambled it to make a type specimen book. It has survived not only five` +
      `centuries, but also the leap into electronic typesetting, remaining` +
      `essentially unchanged. It was popularised in the 1960s with the release` +
      `of Letraset sheets containing Lorem Ipsum passages, and more recently` +
      `with desktop publishing software like Aldus PageMaker including versions` +
      `of Lorem Ipsum.`;
    const logViewer = new LogViewer({ logLines: [line] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders a multi-line log', () => {
    const line = `Lorem Ipsum is simply dummy text of the printing and typesetting
    industry. Lorem Ipsum has been the industry's standard dummy text ever
    since the 1500s, when an unknown printer took a galley of type and
    scrambled it to make a type specimen book. It has survived not only five
    centuries, but also the leap into electronic typesetting, remaining
    essentially unchanged. It was popularised in the 1960s with the release
    of Letraset sheets containing Lorem Ipsum passages, and more recently
    with desktop publishing software like Aldus PageMaker including versions
    of Lorem Ipsum.`;
    const logViewer = new LogViewer({ logLines: line.split('\n') });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('linkifies standalone urls', () => {
    const logLines = ['this string: http://path.com is a url'];
    const logViewer = new LogViewer({ logLines });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('linkifies standalone https urls', () => {
    const logLines = ['this string: https://path.com is a url'];
    const logViewer = new LogViewer({ logLines });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('linkifies substring urls', () => {
    const logLines = ['this string:http://path.com is a url'];
    const logViewer = new LogViewer({ logLines });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('does not linkify non http/https urls', () => {
    const logLines = ['this string: gs://path is a GCS path'];
    const logViewer = new LogViewer({ logLines });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('scrolls to end after rendering', () => {
    const spy = jest.spyOn(LogViewer.prototype as any, '_scrollToEnd');
    const logs = 'this string: gs://path is a GCS path';
    const tree = mount(<LogViewer logLines={[logs]} />);
    tree.instance().componentDidUpdate!({}, {});
    expect(spy).toHaveBeenCalled();
  });

  it('renders a row with given index as line number', () => {
    const logViewer = new LogViewer({ logLines: ['line1', 'line2'] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders a row with error', () => {
    const logViewer = new LogViewer({ logLines: ['line1 with error', 'line2'] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders a row with upper case error', () => {
    const logViewer = new LogViewer({ logLines: ['line1 with ERROR', 'line2'] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders a row with error word as substring', () => {
    const logViewer = new LogViewer({ logLines: ['line1 with errorWord', 'line2'] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders a row with warning', () => {
    const logViewer = new LogViewer({ logLines: ['line1 with warning', 'line2'] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders a row with warn', () => {
    const logViewer = new LogViewer({ logLines: ['line1 with warn', 'line2'] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders a row with upper case warning', () => {
    const logViewer = new LogViewer({ logLines: ['line1 with WARNING', 'line2'] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });

  it('renders a row with warning word as substring', () => {
    const logViewer = new LogViewer({ logLines: ['line1 with warning:something', 'line2'] });
    const tree = mount((logViewer as any)._rowRenderer({ index: 0 })).getDOMNode();
    expect(tree).toMatchSnapshot();
  });
});
