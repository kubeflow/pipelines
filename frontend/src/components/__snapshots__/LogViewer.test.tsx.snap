// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogViewer does not linkify non http/https urls 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        this string: gs://path is a GCS path
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer linkifies standalone https urls 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        this string: 
      </span>
    </span>
    <span>
      <a
        class="a"
        href="https://path.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://path.com
      </a>
    </span>
    <span>
      <span>
         is a url
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer linkifies standalone urls 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        this string: 
      </span>
    </span>
    <span>
      <a
        class="a"
        href="http://path.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        http://path.com
      </a>
    </span>
    <span>
      <span>
         is a url
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer linkifies substring urls 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        this string:
      </span>
    </span>
    <span>
      <a
        class="a"
        href="http://path.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        http://path.com
      </a>
    </span>
    <span>
      <span>
         is a url
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a multi-line log 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        Lorem Ipsum is simply dummy text of the printing and typesetting
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a row with error 1`] = `
<div
  class="line"
>
  <span
    class="number"
    style="background-color: rgb(112, 0, 0); color: white;"
  >
    1
  </span>
  <span
    class="line"
    style="background-color: rgb(112, 0, 0); color: white;"
  >
    <span>
      <span>
        line1 with error
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a row with error word as substring 1`] = `
<div
  class="line"
>
  <span
    class="number"
    style="background-color: rgb(112, 0, 0); color: white;"
  >
    1
  </span>
  <span
    class="line"
    style="background-color: rgb(112, 0, 0); color: white;"
  >
    <span>
      <span>
        line1 with errorWord
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a row with given index as line number 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        line1
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a row with upper case error 1`] = `
<div
  class="line"
>
  <span
    class="number"
    style="background-color: rgb(112, 0, 0); color: white;"
  >
    1
  </span>
  <span
    class="line"
    style="background-color: rgb(112, 0, 0); color: white;"
  >
    <span>
      <span>
        line1 with ERROR
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a row with upper case warning 1`] = `
<div
  class="line"
>
  <span
    class="number"
    style="background-color: rgb(84, 84, 0); color: white;"
  >
    1
  </span>
  <span
    class="line"
    style="background-color: rgb(84, 84, 0); color: white;"
  >
    <span>
      <span>
        line1 with WARNING
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a row with warn 1`] = `
<div
  class="line"
>
  <span
    class="number"
    style="background-color: rgb(84, 84, 0); color: white;"
  >
    1
  </span>
  <span
    class="line"
    style="background-color: rgb(84, 84, 0); color: white;"
  >
    <span>
      <span>
        line1 with warn
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a row with warning 1`] = `
<div
  class="line"
>
  <span
    class="number"
    style="background-color: rgb(84, 84, 0); color: white;"
  >
    1
  </span>
  <span
    class="line"
    style="background-color: rgb(84, 84, 0); color: white;"
  >
    <span>
      <span>
        line1 with warning
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders a row with warning word as substring 1`] = `
<div
  class="line"
>
  <span
    class="number"
    style="background-color: rgb(84, 84, 0); color: white;"
  >
    1
  </span>
  <span
    class="line"
    style="background-color: rgb(84, 84, 0); color: white;"
  >
    <span>
      <span>
        line1 with warning:something
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders an empty container when no logs passed 1`] = `
<AutoSizer
  disableHeight={false}
  disableWidth={false}
  onResize={[Function]}
  style={Object {}}
>
  <Component />
</AutoSizer>
`;

exports[`LogViewer renders one log line 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        first line
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders one long line without breaking 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        Lorem Ipsum is simply dummy text of the printing and typesettingindustry. Lorem Ipsum has been the industry's standard dummy text eversince the 1500s, when an unknown printer took a galley of type andscrambled it to make a type specimen book. It has survived not only fivecenturies, but also the leap into electronic typesetting, remainingessentially unchanged. It was popularised in the 1960s with the releaseof Letraset sheets containing Lorem Ipsum passages, and more recentlywith desktop publishing software like Aldus PageMaker including versionsof Lorem Ipsum.
      </span>
    </span>
  </span>
</div>
`;

exports[`LogViewer renders two log lines 1`] = `
<div
  class="line"
>
  <span
    class="number"
  >
    1
  </span>
  <span
    class="line"
  >
    <span>
      <span>
        first line
      </span>
    </span>
  </span>
</div>
`;
